pkg/executor/local/executor.go

```diff
@@ -0,0 +1,418 @@
+package local
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"runtime/debug"
+
+	"github.com/google/uuid"
+
+	dockertypes "github.com/docker/docker/api/types"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/network"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/filecoin-project/bacalhau/pkg/capacitymanager"
+	"github.com/filecoin-project/bacalhau/pkg/config"
+	"github.com/filecoin-project/bacalhau/pkg/docker"
+	"github.com/filecoin-project/bacalhau/pkg/executor"
+	"github.com/filecoin-project/bacalhau/pkg/storage"
+	"github.com/filecoin-project/bacalhau/pkg/storage/util"
+	"github.com/filecoin-project/bacalhau/pkg/system"
+	"github.com/rs/zerolog/log"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const NanoCPUCoefficient = 1000000000
+
+type Local struct {
+	ID string
+
+	jobID string
+
+	ResultsDir string
+
+	StorageProviders map[storage.StorageSourceType]storage.StorageProvider
+
+	Client *dockerclient.Client
+}
+
+// var jobID string
+
+func NewExecutor(
+	cm *system.CleanupManager,
+	id string,
+	storageProviders map[storage.StorageSourceType]storage.StorageProvider,
+) (*Local, error) {
+	dockerClient, err := docker.NewDockerClient()
+	jobID := genID()
+	if err != nil {
+		return nil, err
+	}
+
+	dir, err := ioutil.TempDir("", "bacalhau-docker-executor")
+	if err != nil {
+		return nil, err
+	}
+
+	de := &Local{
+		ID:               id,
+		jobID:            jobID,
+		ResultsDir:       dir,
+		StorageProviders: storageProviders,
+		Client:           dockerClient,
+	}
+
+	cm.RegisterCallback(func() error {
+		de.cleanupAll()
+		return nil
+	})
+
+	return de, nil
+}
+
+// write a function for interactive mode
+
+func (e *Local) getStorageProvider(ctx context.Context, engine storage.StorageSourceType) (storage.StorageProvider, error) {
+	return util.GetStorageProvider(ctx, engine, e.StorageProviders)
+}
+
+// IsInstalled checks if docker itself is installed.
+func (e *Local) IsInstalled(ctx context.Context) (bool, error) {
+	return docker.IsInstalled(e.Client), nil
+}
+
+func (e *Local) HasStorageLocally(ctx context.Context, volume storage.StorageSpec) (bool, error) {
+	ctx, span := newSpan(ctx, "HasStorageLocally")
+	defer span.End()
+
+	s, err := e.getStorageProvider(ctx, volume.Engine)
+	if err != nil {
+		return false, err
+	}
+
+	return s.HasStorageLocally(ctx, volume)
+}
+
+func (e *Local) GetVolumeSize(ctx context.Context, volume storage.StorageSpec) (uint64, error) {
+	storageProvider, err := e.getStorageProvider(ctx, volume.Engine)
+	if err != nil {
+		return 0, err
+	}
+	return storageProvider.GetVolumeSize(ctx, volume)
+}
+
+// TODO: #289 Clean up RunJob
+// nolint:funlen,gocyclo // will clean up
+func (e *Local) RunJobLocally(ctx context.Context, jobSpec executor.JobSpec) (string, error) {
```

Comment:
I'm definitely not in favour of a picky linter but there may be room to shorten this func a bit?